src/one/util/ideaplugin/screenshoter/ImageBuilder.java

```diff
@@ -0,0 +1,146 @@
+
+    ImageBuilder(@NotNull Editor editor) {
+        this.editor = editor;
+    }
+
+    @NotNull
+    BufferedImage createImage() {
+        SelectionModel selectionModel = editor.getSelectionModel();
+        int start = selectionModel.getSelectionStart();
+        int end = selectionModel.getSelectionEnd();
+
+        selectionModel.setSelection(0, 0);
+
+        CaretModel caretModel = editor.getCaretModel();
+        
+        Document document = editor.getDocument();
+        int offset = caretModel.getOffset();
+        
+        CopyImageOptionsProvider.State options = CopyImageOptionsProvider.getInstance(editor.getProject()).getState();
+        if (options.myRemoveCaret) {
+            caretModel.moveToOffset(start == 0 ? document.getLineEndOffset(document.getLineCount() - 1) : 0);
+        }
+
+        String text = document.getText(new TextRange(start, end));
+
+        double scale = options.myScale;
+        JComponent contentComponent = editor.getContentComponent();
+        Graphics2D contentGraphics = (Graphics2D) contentComponent.getGraphics();
+        AffineTransform currentTransform = contentGraphics.getTransform();
+        AffineTransform newTransform = new AffineTransform(currentTransform);
+        newTransform.scale(scale, scale);
+        // To flush glyph cache
+        paint(contentComponent, newTransform, 1, 1);
+        Rectangle2D r = new Rectangle2D.Double();
+
+        for (int i = start; i <= end; i++) {
+            if (options.myChopIndentation &&
+                    EMPTY_SUFFIX.matcher(text.substring(0, Math.min(i - start + 1, text.length()))).find()) {
+                continue;
+            }
+            VisualPosition pos = editor.offsetToVisualPosition(i);
+            Point2D point = getPoint(editor, pos);
+            includePoint(r, point);
+            includePoint(r, new Point2D.Double(point.getX(), point.getY() + editor.getLineHeight()));
+        }
+
+        newTransform.translate(-r.getX(), -r.getY());
+        BufferedImage editorImage = paint(contentComponent, newTransform,
+                (int) (r.getWidth() * scale), (int) (r.getHeight() * scale));
+        BufferedImage image = padding(options, editor, editorImage);
+
+        selectionModel.setSelection(start, end);
+        caretModel.moveToOffset(offset);
+        return image;
+    }
+
+
+    private BufferedImage padding(CopyImageOptionsProvider.State options,
+                                  @NotNull Editor editor,
+                                  @NotNull BufferedImage image) {
+        int padding = options.myPadding;
+        if (padding == 0) return image;
+
+        int sizeX = image.getWidth() + padding * 2;
+        int sizeY = image.getHeight() + padding * 2;
+        //noinspection UndesirableClassUsage / already scaled
+        BufferedImage newImage = new BufferedImage(sizeX, sizeY, BufferedImage.TYPE_INT_RGB);
+        
+        Color backgroundColor = getBackgroundColor(editor, false);
```

Comment:
Just curious: how it would behave if editor background image is used?

Reply:
it uses background color from project's color schema and knows nothing about background image but I'd say that it is not a big problem